### PR TITLE
Fix some bugs in TypeCoercion rule

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/3390

 # Rationale for this change

This PR fixes a number of bugs that I discovered when running a popular SQL benchmark. It does not fix all of them.

# What changes are included in this PR?

- Fix schema bug that affected queries with more than 2 inputs (union, intersect, etc)
- Add type coercion for `BETWEEN`
- Add workaround for a bug with `+ INTERVAL`
- Add coercion support for `Date32` and `Date64`

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/optimizer/src/type_coercion.rs
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:57]: 👍

File: datafusion/optimizer/src/type_coercion.rs
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:104]: I don't think it makes sense to cast an interval to a specific point in time -- aka I think DataFusion's coerce_types is incorrect

File: datafusion/optimizer/src/type_coercion.rs
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:131]: Why the empty error message? That seems like it may be confusing

File: datafusion/optimizer/src/type_coercion.rs
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:288]: ❤️

File: datafusion/expr/src/binary_rule.rs
Comment (alamb) [inline datafusion/expr/src/binary_rule.rs:519]: 🤔  I think coercing from Date32 -> Date64 is probably fine, but going the other way loses precision, right?

**Review by alamb — APPROVED** [review]
👍  I had a small question about casting Date32 -> date64 but otherwise this looks like a nice improvement to me

**Comment by liukun4515** [comment]
I will review tomorrow, please hold it

File: datafusion/expr/src/binary_rule.rs
Comment (andygrove) [reply datafusion/expr/src/binary_rule.rs:519]: My understanding of this code is that it is finding the common type that both lhs_type and rhs_type can safely be cast to, so any combination of `Date32` and `Date64` can be cast to `Date64`

File: datafusion/optimizer/src/type_coercion.rs
Comment (andygrove) [reply datafusion/optimizer/src/type_coercion.rs]: oops .. forgot to go back and finish this. I have pushed a fix.

File: datafusion/optimizer/src/type_coercion.rs
Comment (andygrove) [reply datafusion/optimizer/src/type_coercion.rs:104]: I am tracking this as part of https://github.com/apache/arrow-datafusion/issues/3419

File: datafusion/expr/src/binary_rule.rs
Comment (alamb) [reply datafusion/expr/src/binary_rule.rs:519]: Yes, you are correct. Sorry about that

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
> I will review tomorrow, please hold it



@liukun4515  do you still want us to hold this PR for your review?

**Comment by andygrove** [comment]
I'd prefer that we get these bug fixes merged to unblock some other items. Arrow is still following the "Commit-Then-Review" (C-T-R) policy:

```
A policy governing code changes which permits developers to make changes at will, with the possibility of being retroactively [vetoed](http://www.apache.org/foundation/glossary.html#Veto)
```
In other words, we can continue with the review after merge, and revert the PR if it proves to be problematic.

**Comment by alamb** [comment]
I agree -- let's merge this PR and we can either revert it or address comments in a follow on PR if/when provided by @liukun4515 

Thanks all!

**Pull request merged**
